pallets/wnodl/src/lib.rs

```diff
@@ -0,0 +1,317 @@
+        /// parameters. [account's address on Nodle chain, amount of Nodl fund settled, Transaction hash on Ethereum main-net, reporting oracle's address]
+        WrappingSettled(T::AccountId, BalanceOf<T>, EthTxHash),
+
+        /// Wrapping limits is set
+        /// parameters. [minimum amount, maximum amount]
+        LimitSet(BalanceOf<T>, BalanceOf<T>),
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        /// The balance of the account is not sufficient for the requested transaction
+        BalanceNotEnough,
+        /// The amount of fund to wrap should be between the pre-specified limits for the pallet.
+        FundNotWithinLimits,
+        /// We do not expect this error ever happen. But if it happened we would not allow it to mess with the correctness of the storage.
+        BalanceOverflow,
+        /// The customer is not known, whitelisted for this operation.
+        NotEligible,
+        /// Settlment is only possible for an amount equal or smaller than an initiated wrapping for a known customer
+        InvalidSettle,
+        /// Min must be less than max whern setting limits
+        InvalidLimits,
+    }
+
+    #[pallet::genesis_config]
+    pub struct GenesisConfig<T: Config> {
+        pub min_wrapping: BalanceOf<T>,
+        pub max_wrapping: BalanceOf<T>,
+    }
+
+    #[cfg(feature = "std")]
+    impl<T: Config> Default for GenesisConfig<T> {
+        fn default() -> Self {
+            Self {
+                min_wrapping: Zero::zero(),
+                max_wrapping: Bounded::max_value(),
+            }
+        }
+    }
+
+    #[pallet::genesis_build]
+    impl<T: Config> GenesisBuild<T> for GenesisConfig<T> {
+        fn build(&self) {
+            CurrentMin::<T>::put(self.min_wrapping);
+            CurrentMax::<T>::put(self.max_wrapping);
+        }
+    }
+
+    #[cfg(feature = "std")]
+    impl<T: Config> GenesisConfig<T> {
+        /// Direct implementation of `GenesisBuild::build_storage`.
+        ///
+        /// Kept in order not to break dependency.
+        pub fn build_storage(&self) -> Result<sp_runtime::Storage, String> {
+            <Self as GenesisBuild<T>>::build_storage(self)
+        }
+
+        /// Direct implementation of `GenesisBuild::assimilate_storage`.
+        ///
+        /// Kept in order not to break dependency.
+        pub fn assimilate_storage(&self, storage: &mut sp_runtime::Storage) -> Result<(), String> {
+            <Self as GenesisBuild<T>>::assimilate_storage(self, storage)
+        }
+    }
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        /// Initiate wrapping an amount of Nodl into wnodl on Ethereum
+        #[pallet::weight(10_000 + T::DbWeight::get().writes(1))]
+        pub fn initiate_wrapping(
+            origin: OriginFor<T>,
+            amount: BalanceOf<T>,
+            eth_dest: EthAddress,
+        ) -> DispatchResult {
+            let who = ensure_signed(origin)?;
+
+            #[cfg(feature = "runtime-benchmarks")]
```

Comment:
We need to whitelist some account ids for the purpose of benchmarking. This is an easy solution for that which is not deviating from the original execution sensibly. I would welcome a better solution for adding some accounts as the known members for runtime benchmarking.

Reply:
I hope pallet-grants benchmark is good reference in this context

https://github.com/NodleCode/chain/blob/master/pallets/grants/src/benchmarking.rs

Reply:
The technique that is used in pallet-grants doesn't seem to be useful in our case. The reason is, within the pallet-wnodl the only information that is known about the known customers is that it implements `Contains`:
`type KnownCustomers: Contains<Self::AccountId>`
Which means we cannot add to the known customers from within. Does it make sense?

Reply:
As part of mock::new_test_ext() `T::Oracles` & `T::KnownCustomers` storage
can be initialised with known address for Oracles & knowncustomers, & bechmarking 
for `reject() & settle()` can be done on top of it. 
in this way build time diversity `#[cfg(feature = "runtime-benchmarks")]` can be avoided.

Reply:
The storage that we set in mock::new_test_ext() would only work when testing the benchmark tests in the test mode. The benchmarking itself is written in a generic way so it could get the actual Runtime when we run it like:
`./target/release/nodle-chain benchmark --chain dev ...`
It would actually use our main-net Runtime configuration. At that point the test storage will be not available.

Reply:
You are right ...

For production or final benchmarking weight computation, chainspec is picked up based on '--chain dev' config params.

So for chain spec `dev`, here storage of our interest is ...

from `pallet_membership`

| Interface | RT instance | pallet instance |
| --- | --- | --- |
| T::Oracles | WnodlOracleMembership | pallet_membership::<Instance6> |
| T::KnownCustomers | KnownCustomerMembership | pallet_membership::<Instance7> |

which get initialised as part of genesis here 
...
    
![](https://i.imgur.com/6w0a41g.png)

I hope for pallet `pallet-wnodl` below storage gets introduced only to handle "runtime-benchmarks"

```
WhitelistedCallers
```

![](https://i.imgur.com/lQF9pwx.png)

In my view its not required, For crate unittest benchmarking context, we can reuse the mock initialised config ...
    
```
pub const KNOWN_CUSTOMERS: [u64; 4] = [4, 5, 7, 1];
pub const TRUSTED_ORACLES: [u64; 2] = [1, 3];

pub struct Oracles;
impl Contains<u64> for Oracles {
    fn contains(t: &u64) -> bool {
        return TRUSTED_ORACLES.contains(t);
    }
}

pub struct KnownCustomers;
impl Contains<u64> for KnownCustomers {
    fn contains(t: &u64) -> bool {
        return KNOWN_CUSTOMERS.contains(t);
    }
}
 
impl pallet_wnodl::Config for Test {
    type Event = Event;
    type Currency = Balances;
    type Oracles = Oracles;
    type KnownCustomers = KnownCustomers;
    type Reserve = Reserve;
    type WeightInfo = ();
}

```

Ideally this mock config is available for benchmarking context also, then why we need extra storage `WhitelistedCallers` ?

Reply:
There are 2 points here:
1. When doing benchmarking for production we should be able to do the test on different chain configurations. Technically the results for "--chain dev" may not be necessarily apply-able to other configurations. 
2. The `whitelistedCallers` storage only exists for production and replaces the similar storage that we would have for KnownCurstomers. 
All in all what I am saying is that this is a balanced solution that is good enough for the purpose.